### PR TITLE
feat(prompts): message placeholders in js sdk

### PR DESCRIPTION
**Pull request by nimarb**
> [!IMPORTANT]
> Adds message placeholders to Langfuse JS SDK for dynamic prompt customization, updates OpenAPI specs, and enhances test coverage.
> 
>   - **Behavior**:
>     - Adds `ChatMessageWithPlaceholders` type in `promptClients.ts` for dynamic message insertion.
>     - Updates `compile()` and `getLangchainPrompt()` in `ChatPromptClient` to handle placeholders.
>     - Supports placeholders in `createPrompt()` in `index.ts`.
>   - **OpenAPI**:
>     - Adds placeholder support in `openapi-client.yaml` and `openapi-server.yaml`.
>     - Updates `CreateScoreRequest` in `client.ts` with optional fields.
>   - **Testing**:
>     - Adds tests in `langfuse.prompts.spec.ts` for placeholder functionality and edge cases.
>   - **Misc**:
>     - Renames `scoreGet` to `scoreV2Get` in `langfuse-public-api.spec.ts`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=langfuse%2Flangfuse-js&utm_source=github&utm_medium=referral)<sup> for e37270818a8b6e2d5432fc065d187a52f4c51ba0. You can [customize](https://app.ellipsis.dev/langfuse/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

----

## Greptile Summary

**Disclaimer**: Experimental PR review
---
Implements support for message placeholders in chat prompts within the Langfuse JS SDK, allowing dynamic insertion of message groups like conversation history into prompts.

- Enhanced `langfuse-core/src/prompts/promptClients.ts` with new `ChatMessageWithPlaceholders` type and updated compilation logic for placeholder replacement
- Added placeholder support in OpenAPI specs (`openapi-client.yaml` and `openapi-server.yaml`) with new message type definitions
- Updated scoring endpoint in `langfuse-core/src/openapi/client.ts` with optional fields for `sessionId`, `datasetRunId`, and metadata
- Added comprehensive test coverage in `langfuse-core/test/langfuse.prompts.spec.ts` for placeholder functionality and edge cases

### CODE REVIEW

File: langfuse-core/src/index.ts [hunk 1]
```diff
@@ -1476,15 +1477,32 @@ export abstract class LangfuseCore extends LangfuseCoreStateless {
...
     const labels = body.labels ?? [];
 
     const promptResponse =
       body.type === "chat" // necessary to get types right here
         ? await this.createPromptStateless({
             ...body,
+            prompt: body.prompt.map((item) => {
+              if ("type" in item && item.type === "placeholder") {
+                return { type: "placeholder", name: (item as PlaceholderMessage).name };
+              } else if ("type" in item && item.type === "chatmessage") {
```
Comment (hassiebp) [inline langfuse-core/src/index.ts:1495]: when will this code block execute, will we ever get type === 'chatmessage' from the API?

File: langfuse-core/src/prompts/promptClients.ts [hunk 2]
```diff
@@ -147,16 +157,49 @@ export class TextPromptClient extends BasePromptClient {
...
-    this.prompt = prompt.prompt;
+  constructor(prompt: ChatPromptCompat, isFallback = false) {
+    // Convert ChatMessages to ChatMessageWithPlaceholders for backward compatibility
+    const normalizedPrompt = prompt.prompt.map((item): ChatMessageWithPlaceholders => {
+      if ("type" in item) {
+        // Already has type field (new format)
+        return item as ChatMessageWithPlaceholders;
+      } else {
+        // Plain ChatMessage (legacy format) - add type field
+        return { type: "chatmessage" as const, ...item } as ChatMessageWithPlaceholders;
```
Comment (hassiebp) [inline langfuse-core/src/prompts/promptClients.ts:170]: we probably should use an enum here for `chatmessage` to make sure we don't run into any issues with typos

File: langfuse-core/src/prompts/promptClients.ts [hunk 3]
```diff
@@ -182,10 +225,32 @@ export class ChatPromptClient extends BasePromptClient {
      * ```
      * @returns {ChatMessage[]} Chat messages with variables that can be plugged into Langchain's ChatPromptTemplate.
      */
-    return this.prompt.map((chatMessage) => ({
-      ...chatMessage,
-      content: this._transformToLangchainVariables(chatMessage.content),
-    }));
+    return this.prompt
+      .filter((item) => "role" in item && "content" in item)
```
Comment (hassiebp) [inline langfuse-core/src/prompts/promptClients.ts:229]: same as in python sdk, we should probably keep all passed in placeholder messages

File: langfuse-core/src/prompts/promptClients.ts [hunk 4]
```diff
@@ -143,26 +153,138 @@ export class TextPromptClient extends BasePromptClient {
...
+      config: this.config,
+    });
+  }
 }
 
 export class ChatPromptClient extends BasePromptClient {
   public readonly promptResponse: ChatPrompt;
-  public readonly prompt: ChatMessage[];
+  private rawPrompt!: ChatMessageWithPlaceholders[];
+  private placeholderFillIns: Record<string, ChatMessage[]> = {};
```
Comment (hassiebp) [inline langfuse-core/src/prompts/promptClients.ts:174]: A Map might be a better choice here in order to not run into any issues with key naming where keys are reserved as e.g. Object.prototype

**Review by hassiebp — APPROVED** [review]
